packages/react-noop-renderer/src/createReactNoop.js

```diff
@@ -873,41 +868,52 @@ function createReactNoop(reconciler: Function, useMutation: boolean) {
 
     interactiveUpdates: NoopRenderer.interactiveUpdates,
 
-    // maybe this should exist only in the test file
-    act(callback: () => void): Thenable {
+    act(callback: () => void | Promise<void>) {
       // note: keep these warning messages in sync with
       // ReactTestRenderer.js and ReactTestUtils.js
-      let result = NoopRenderer.batchedUpdates(callback);
-      if (__DEV__) {
-        if (result !== undefined) {
-          let addendum;
-          if (result !== null && typeof result.then === 'function') {
-            addendum =
-              "\n\nIt looks like you wrote ReactNoop.act(async () => ...) or returned a Promise from it's callback. " +
-              'Putting asynchronous logic inside ReactNoop.act(...) is not supported.\n';
-          } else {
-            addendum = ' You returned: ' + result;
-          }
-          warningWithoutStack(
-            false,
-            'The callback passed to ReactNoop.act(...) function must not return anything.%s',
-            addendum,
-          );
+      const result = NoopRenderer.actedUpdates(callback);
+      if (
+        result !== null &&
+        typeof result === 'object' &&
+        typeof result.then === 'function'
+      ) {
+        let called = false;
+        if (__DEV__) {
+          setTimeout(() => {
+            if (!called) {
+              warningWithoutStack(
+                null,
+                'You called act() without awaiting its result. ' +
+                  'This could lead to unexpected testing behaviour, interleaving multiple act ' +
+                  'calls and mixing their scopes. You should - await act(async () => ...);',
+                // todo - a better warning here. open to suggestions.
+              );
+            }
+          }, 0);
```

Comment:
Using a timeout isn't ideal because it only works if the timers run. Let's figure out something better. At the very least, we should use a microtask instead of a timer event.

(In retrospect, maybe what we should have done is wait to call the callback until the thenable is awaited. Then `act` wouldn't do anything _unless_ you await the promise. But I don't think there's a way to change that without breaking everyone's existing tests.)

Reply:
Yeah I think if you make this a microtask (Promise.then()) instead of `setTimeout` then it's good enough. Jest will still fail even if nothing is awaited.

Reply:
Welp, never mind, that only works because of the way we set up Jest in the React repo.

Reply:
Still, microtask is probably our best option. We can looking into having Jest always wait a microtask at the end each test.

Reply:
the problem with the microtask is that it resolves before any further created promises resolve, triggering the warning every time.

Reply:
How about two microtasks? :D

https://github.com/facebook/react/blob/a477f6d2fa056fdb54bf8a58037f7c225b4fb997/packages/react-noop-renderer/src/createReactNoop.js#L841-L848